{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MSLRVRQIDRREWLLAQTATECQRHGREATLEYPTRQGMWVRLSDAEKRWSAWIKPGDWLEHVSPALAGAAVSAGAEHLVVPWLAATERPFELPVPHLSCRRLCVENPVPGSALPEGKLLHIMSDRGGLWFEHLPELPAVGGGRPKMLRWPLRFVIGSSDTQRSLLGRIGIGDVLLIRTSRAEVYCYAKKLGHFNRVEGGIIVETLDIQHIEEENNTTETAETLPGLNQLPVKLEFVLYRKNVTLAELEAMGQQQLLSLPTNAELNVEIMANGVLLGNGELVQMNDTLGVEIHEWLSESGNGE', 'MSSNKTEKPTKKRLEDSAKKGQSFKSKDLIIACLTLGGIAYLVSYGSFNEFMGIIKIIIADNFDQSMADYSLAVFGIGLKYLIPFMLLCLVCSALPALLQAGFVLATEALKPNLSALNPVEGAKKLFSMRTVKDTVKTLLYLSSFVVAAIICWKKYKVEIFSQLNGNIVGIAVIWRELLLALVLTCLACALIVLLLDAIAEYFLTMKDMKMDKEEVKREMKEQEGNPEVKSKRREVHMEILSEQVKSDIENSRLIVANPTHITIGIYFKPELMPIPMISVYETNQRALAVRAYAEKVGVPVIVDIKLARSLFKTHRRYDLVSLEEIDEVLRLLVWLEEVENAGKDVIQPQENEVRH', 'MGIFASAGCGKTMLMHMLIEQTEADVFVIGLIGERGREVTEFVDMLRASHKKEKCVLVFATSDFPSVDRCNAAQLATTVAEYFRDQGKRVVLFIDSMTRYARALRDVALASGERPARRGYPASVFDNLPRLLERPGATSEGSITAFYTVLLESEEEADPMADEIRSILDGHLYLSRKLAGQGHYPAIDVLKSVSRVFGQVTTPTHAEQASAVRKLMTRLEELQLFIDLGEYRPGENIDNDRAMQMRDSLKAWLCQPVAQYSSFDDTLSGMNAFADQN', 'MGDVSAVSSSGNILLPQQDEVGGLSEALKKAVEKHKTEYSGDKKDRDYGDAFVMHKETALPLLLAAWRHGAPAKSEHHNGNVSGLHHNGKSELRIAEKLLKVTAEKSVGLISAEAKVDKSAALLSSKNRPLESVSGKKLSADLKAVESVSEVTDNATGISDDNIKALPGDNKAIAGEGVRKEGAPLARDVAPARMAAANTGKPEDKDHKKVKDVSQLPLQPTTIADLSQLTGGDEKMPLAAQSKPMMTIFPTADGVKGEDSSLTYRFQRWGNDYSVNIQARQAGEFSLIPSNTQVEHRLHDQWQNGNPQRWHLTRDDQQNPQQQQHRQQSGEEDDA', 'MFYALYFEIHHLVASAALGFARVAPIFFFLPFLNSGVLSGAPRNAIIILVALGVWPHALNEAPPFLSVAMIPLVLQEAAVGVMLGCLLSWPFWVMHALGCIIDNQRGATLSSSIDPANGIDTSEMANFLNMFAAVVYLQNGGLVTMVDVLNKSYQLCDPMNECTPSLPPLLTFINQVAQNALVLASPVVLVLLLSEVFLGLLSRFAPQMNAFAISLTVKSGIAVLIMLLYFSPVLPDNVLRLSFQATGLSSWFYERGATHVLE']\n"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "PYTHON MINI-PROJECT FOR MSC. BIOINFORMATICS\n",
    "\n",
    "@author: ALEX KIMANI SG30/PU/36049/20 \n",
    "\n",
    "\"\"\"\n",
    "\n",
    "import random\n",
    "from amino_acids import aa, codons, aa_table   # you may find these useful\n",
    "from load import load_seq\n",
    "dna = load_seq(\"./data/X73525.fa\")\n",
    "\n",
    "def shuffle_string(s):\n",
    "    \"\"\"Shuffles the characters in the input string\n",
    "        NOTE: this is a helper function, you do not\n",
    "        have to modify this in any way \"\"\"\n",
    "    return ''.join(random.sample(s, len(s)))\n",
    "\n",
    "# YOU WILL START YOUR IMPLEMENTATION FROM HERE DOWN ###\n",
    "\n",
    "\n",
    "def get_complement(nucleotide):\n",
    "    \"\"\" Returns the complementary nucleotide\n",
    "\n",
    "        nucleotide: a nucleotide (A, C, G, or T) represented as a string\n",
    "        returns: the complementary nucleotide\n",
    "    >>> get_complement('A')\n",
    "    'T'\n",
    "    >>> get_complement('C')\n",
    "    'G'\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    complement = ''\n",
    "    for i in range(len(nucleotide)):\n",
    "        x = nucleotide[i]\n",
    "\n",
    "        if x == 'A':\n",
    "            var = 'T'\n",
    "        elif x =='T':\n",
    "            var = 'A'\n",
    "        elif x == 'C':\n",
    "            var = 'G'\n",
    "        else:\n",
    "            var = 'C'\n",
    "\n",
    "        complement = complement + var\n",
    "    return complement\n",
    "    \n",
    "    #pass\n",
    "\n",
    "def get_reverse_complement(dna):\n",
    "    \"\"\" Computes the reverse complementary sequence of DNA for the specfied DNA\n",
    "        sequence\n",
    "\n",
    "        dna: a DNA sequence represented as a string\n",
    "        returns: the reverse complementary DNA sequence represented as a string\n",
    "    >>> get_reverse_complement(\"ATGCCCGCTTT\")\n",
    "    'AAAGCGGGCAT'\n",
    "    >>> get_reverse_complement(\"CCGCGTTCA\")\n",
    "    'TGAACGCGG'\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    reverse_dna = dna[::-1]\n",
    "    reverse_complement = get_complement(reverse_dna)\n",
    "    return reverse_complement\n",
    "\n",
    "    #pass\n",
    "\n",
    "\n",
    "def rest_of_ORF(dna):\n",
    "    \"\"\" Takes a DNA sequence that is assumed to begin with a start\n",
    "        codon and returns the sequence up to but not including the\n",
    "        first in frame stop codon.  If there is no in frame stop codon,\n",
    "        returns the whole string.\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        returns: the open reading frame represented as a string\n",
    "    >>> rest_of_ORF(\"ATGTGAA\")\n",
    "    'ATG'\n",
    "    >>> rest_of_ORF(\"ATGAGATAGG\")\n",
    "    'ATGAGA'\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    ORF = ''\n",
    "    x = 0\n",
    "    for x in range(0,len(dna),3):\n",
    "        i = dna[x:x+3]\n",
    "        if dna[x:x+3] == 'TGA':\n",
    "            return ORF\n",
    "        if dna[x:x+3] == 'TAG':\n",
    "            return ORF\n",
    "        if dna[x:x+3] == 'TAA':\n",
    "            return ORF\n",
    "        ORF = ORF + i\n",
    "    return ORF\n",
    "    #pass\n",
    "\n",
    "def find_all_ORFs_oneframe(dna):\n",
    "    \"\"\" Finds all non-nested open reading frames in the given DNA\n",
    "        sequence and returns them as a list.  This function should\n",
    "        only find ORFs that are in the default frame of the sequence\n",
    "        (i.e. they start on indices that are multiples of 3).\n",
    "        By non-nested we mean that if an ORF occurs entirely within\n",
    "        another ORF, it should not be included in the returned list of ORFs.\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        returns: a list of non-nested ORFs\n",
    "    >>> find_all_ORFs_oneframe(\"ATGCATGAATGTAGATAGATGTGCCC\")\n",
    "    ['ATGCATGAATGTAGA', 'ATGTGCCC']\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    ORF2 = []\n",
    "    x = 0\n",
    "    StartCodon = 'ATG'\n",
    "    while x <= len(dna):  \n",
    "        i = dna[x:x+3]\n",
    "        if i == StartCodon:\n",
    "            ORF = rest_of_ORF(dna[x:])\n",
    "            ORF2.append(ORF)\n",
    "            x = x + len(ORF)\n",
    "        x = x+3\n",
    "    return ORF2\n",
    "    #pass\n",
    "\n",
    "\n",
    "def find_all_ORFs(dna):\n",
    "    \"\"\" Finds all non-nested open reading frames in the given DNA sequence in\n",
    "        all 3 possible frames and returns them as a list.  By non-nested we\n",
    "        mean that if an ORF occurs entirely within another ORF and they are\n",
    "        both in the same frame, it should not be included in the returned list\n",
    "        of ORFs.\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        returns: a list of non-nested ORFs\n",
    "\n",
    "    >>> find_all_ORFs(\"ATGCATGAATGTAG\")\n",
    "    ['ATGCATGAATGTAG', 'ATGAATGTAG', 'ATG']\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    ORFtotal = []\n",
    "    ORF1 = find_all_ORFs_oneframe(dna)\n",
    "    ORF2 = find_all_ORFs_oneframe(dna[1:])\n",
    "    ORF3 = find_all_ORFs_oneframe(dna[2:])\n",
    "    ORFtotal.extend(ORF1)\n",
    "    ORFtotal.extend(ORF2)\n",
    "    ORFtotal.extend(ORF3)\n",
    "    return ORFtotal\n",
    "    #pass\n",
    "\n",
    "\n",
    "def find_all_ORFs_both_strands(dna):\n",
    "    \"\"\" Finds all non-nested open reading frames in the given DNA sequence on both\n",
    "        strands.\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        returns: a list of non-nested ORFs\n",
    "    >>> find_all_ORFs_both_strands(\"ATGCGAATGTAGCATCAAA\")\n",
    "    ['ATGCGAATG', 'ATGCTACATTCGCAT']\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    ORF3 = []\n",
    "    ORF1 = find_all_ORFs(dna)\n",
    "    reverse = get_reverse_complement(dna)\n",
    "    ORF2 = find_all_ORFs(reverse)\n",
    "    ORF3.extend(ORF1)\n",
    "    ORF3.extend(ORF2)\n",
    "    return ORF3\n",
    "    #pass\n",
    "\n",
    "\n",
    "def longest_ORF(dna):\n",
    "    \"\"\" Finds the longest ORF on both strands of the specified DNA and returns it\n",
    "        as a string\n",
    "    >>> longest_ORF(\"ATGCGAATGTAGCATCAAA\")\n",
    "    'ATGCTACATTCGCAT'\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    AllORFS = []\n",
    "    x = find_all_ORFs_both_strands(dna)\n",
    "    AllORFS.extend(x)\n",
    "    if not AllORFS == []:\n",
    "        y = max(AllORFS, key=len)\n",
    "        return y\n",
    "    if AllORFS == []:\n",
    "        return ''\n",
    "    #pass\n",
    "\n",
    "\n",
    "def longest_ORF_noncoding(dna, num_trials):\n",
    "    \"\"\" Computes the maximum length of the longest ORF over num_trials shuffles\n",
    "        of the specfied DNA sequence\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        num_trials: the number of random shuffles\n",
    "        returns: the maximum length longest ORF \"\"\"\n",
    "    # TODO: implement this\n",
    "    i = 0\n",
    "    dna2 = dna\n",
    "    PossibleLongest = []\n",
    "    while i <= num_trials:\n",
    "        dna2 = shuffle_string(dna)\n",
    "        longestORF = longest_ORF(dna2)\n",
    "        PossibleLongest.append(longestORF)\n",
    "        i = i + 1\n",
    "    y = max(PossibleLongest, key=len)\n",
    "    return y\n",
    "    #pass\n",
    "\n",
    "\n",
    "def coding_strand_to_AA(dna):\n",
    "    \"\"\" Computes the Protein encoded by a sequence of DNA.  This function\n",
    "        does not check for start and stop codons (it assumes that the input\n",
    "        DNA sequence represents an protein coding region).\n",
    "\n",
    "        dna: a DNA sequence represented as a string\n",
    "        returns: a string containing the sequence of amino acids encoded by the\n",
    "                 the input DNA fragment\n",
    "\n",
    "        >>> coding_strand_to_AA(\"ATGCGA\")\n",
    "        'MR'\n",
    "        >>> coding_strand_to_AA(\"ATGCCCGCTTT\")\n",
    "        'MPA'\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    aminoacidtotal = []\n",
    "    aminoacid = ''\n",
    "    x = 0\n",
    "    dna1 = dna\n",
    "    y = len(dna1) // 3\n",
    "    y = y*3\n",
    "    dna1 = dna1[:y]\n",
    "    for x in range(0,len(dna1),3):\n",
    "        i = str(dna1[x:x+3])\n",
    "        aminoacid = aminoacid + aa_table[i]\n",
    "\n",
    "    return aminoacid\n",
    "    #pass\n",
    "\n",
    "\n",
    "def gene_finder(dna):\n",
    "    \"\"\" Returns the amino acid sequences that are likely coded by the specified dna\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        returns: a list of all amino acid sequences coded by the sequence dna.\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    threshold = len(longest_ORF_noncoding(dna, 1500))\n",
    "    aminoacids = []\n",
    "    Allorfs = find_all_ORFs(dna)\n",
    "    x = 0\n",
    "    for x in range((len(Allorfs))):\n",
    "        i = Allorfs[x]\n",
    "        if len(i) > threshold:\n",
    "            aminoacids.append(coding_strand_to_AA(i))\n",
    "    return aminoacids\n",
    "    #pass\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    import doctest\n",
    "    doctest.testmod()\n",
    "print (gene_finder(dna))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
